recirq/quantum_chess/swap_updater.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2021 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of the swap update algorithm described in the paper 'A Dynamic
+Look-Ahead Heuristic for the Qubit Mapping Problem of NISQ Computers'
+(https://ieeexplore.ieee.org/abstract/document/8976109).
+
+This transforms circuits by adding additional SWAP gates to ensure that all operations are on adjacent qubits.
+"""
+from collections import defaultdict
+from typing import Callable, Dict, Generator, Iterable, List, Optional, Tuple
+
+import cirq
+
+import recirq.quantum_chess.mcpe_utils as mcpe
+
+
+def _satisfies_adjacency(gate: cirq.Operation) -> bool:
+    """Returns true iff the input gate operates on adjacent qubits.
+
+    Requires the input to be either a binary or unary operation on GridQubits.
+    """
+    assert len(gate.qubits) <= 2
+    if len(gate.qubits) < 2:
+        return True
+    q1, q2 = gate.qubits
+    return q1.is_adjacent(q2)
+
+
+def generate_decomposed_swap(
+        q1: cirq.Qid, q2: cirq.Qid) -> Generator[cirq.Operation, None, None]:
+    """Generates a SWAP operation using sqrt-iswap gates."""
```

Comment:
Ouch, good catch. I switched it to use optimize_for_sycamore() for simplicity but LMK if you feel strongly about using the conversion function directly. Or maybe it would be better/simpler to take an optional PointOptimizer instead of a swap_factory callable?

Reply:
Calling optimized_for_sycamore might be less efficient, but it's probably fine for now.